splink/blocking_rules_library.py

```diff
@@ -13,49 +17,131 @@ def __init__(
         """Represents an exact match blocking rule.
 
         Args:
-            col_name (str): Input column name
+            col_name (str): Input column name, or a str represent a sql
+                statement you'd like to match on. For example, `surname` or
+                `"substr(surname,1,2)"` are both valid.
             salting_partitions (optional, int): Whether to add salting
                 to the blocking rule. More information on salting can
-                be found within the docs. Salting is only valid for Spark.
+                be found within the docs. Salting is currently only valid
+                for Spark.
         Examples:
             === ":simple-duckdb: DuckDB"
                 Simple Exact match level
                 ``` python
                 import splink.duckdb.blocking_rule_library as brl
                 brl.exact_match_rule("name")
+
+                sql = "substr(surname,1,2)"
+                brl.exact_match_rule(sql)
                 ```
             === ":simple-apachespark: Spark"
                 Simple Exact match level
                 ``` python
                 import splink.spark.blocking_rule_library as brl
                 brl.exact_match_rule("name", salting_partitions=1)
+
+                sql = "substr(surname,1,2)"
+                brl.exact_match_rule(sql)
                 ```
             === ":simple-amazonaws: Athena"
                 Simple Exact match level
                 ``` python
                 import splink.athena.blocking_rule_library as brl
                 brl.exact_match_rule("name")
+
+                sql = "substr(surname,1,2)"
+                brl.exact_match_rule(sql)
                 ```
             === ":simple-sqlite: SQLite"
                 Simple Exact match level
                 ``` python
                 import splink.sqlite.blocking_rule_library as brl
                 brl.exact_match_rule("name")
+
+                sql = "substr(surname,1,2)"
+                brl.exact_match_rule(sql)
                 ```
             === "PostgreSQL"
                 Simple Exact match level
                 ``` python
                 import splink.postgres.blocking_rule_library as brl
                 brl.exact_match_rule("name")
+
+                sql = "substr(surname,1,2)"
+                brl.exact_match_rule(sql)
                 ```
         """
 
-        col = InputColumn(col_name, sql_dialect=self.sql_dialect)
-        col = col.name()
-        blocking_rule = f"l.{col} = r.{col}"
+        syntax_tree = sqlglot.parse_one(col_name)
+        l_col = add_quotes_and_table_prefix(syntax_tree, "l").sql()
+        r_col = add_quotes_and_table_prefix(syntax_tree, "r").sql()
+
+        blocking_rule = f"{l_col} = {r_col}"
         self._description = "Exact match"
 
         super().__init__(
             blocking_rule,
             salting_partitions=salting_partitions,
         )
+
+
+def block_on_columns(
```

Comment:
Small naming thing - would simply `block_on` be simpler?

I guess this ties into a larger question actually. Do we want to be steering people to use this function as well as `exact_match_rule`? In this PR I think you are assuming that single column rules will use `exact_match_rule` and multi-column ones use this function (which makes the `_columns` fir better)

As alternative, and I think my preference, would be that this function becomes the default and we leave `exact_match_rule` to work purely in the background (and existing user pipelines). From playing around with the function, it appears to work for single column rules so I would be tempted to pivot to it entirely.

Also, I think we had chatted about it previously, but I think it would be good to have a parameter to specify the kind of join between columns (i.e. "and"/"or" with an "and" default). However, happy to leave this until we have scoped out Nick's "or" strategy so we are guiding people only to performant rules.

Sorry, aware this is a bit of an essay - what do you think?

Reply:
Yeah, I agree with virtually everything you've stated here.

* Renaming -> I'm happy w/ `block_on`. The key here is clarity and I think it makes it relatively clear what we're producing. All of our comparison and blocking functions also explicitly tell the user what is contained within when printed to console.
* Yes, I think that's sensible. It makes life easier for anyone trying to use blocking functions within their code.
* For the composition clauses, I think we should wait until we work out a strategy for how we want to split existing rules into more performant versions. There are some functions and methods in which we need to be careful with.

Reply:
It's worth noting the `em_rule` syntax was used within the [inaugural blog post](https://moj-analytical-services.github.io/splink/blog/2022/07/27/splink-updates-july-2023.html#welcome-to-the-splink-blog). Do we want to shift that silently to `block_on`?